runtime/near-vm-runner/src/cache.rs

```diff
@@ -1,37 +1,148 @@
+use wasmer_runtime_core::cache::Artifact;
+use wasmer_runtime_core::load_cache_with;
 
-#[cfg(feature = "no_cache")]
 pub(crate) fn compile_module(
-    _code_hash: Vec<u8>,
     code: &[u8],
     config: &VMConfig,
 ) -> Result<wasmer_runtime::Module, VMError> {
     let prepared_code = prepare::prepare_contract(code, config)?;
     wasmer_runtime::compile(&prepared_code).map_err(|err| err.into_vm_error())
 }
+
+#[derive(Debug, Clone, BorshDeserialize, BorshSerialize)]
+enum ContractCacheKey {
+    Version1 { code_hash: CryptoHash, vm_config_non_crypto_hash: u64, vm_kind: u32 },
+}
+
+#[derive(Debug, Clone, BorshDeserialize, BorshSerialize)]
+enum CacheRecord {
+    Error(VMError),
+    Code(Vec<u8>),
+}
+
+fn get_key(code_hash: &[u8], code: &[u8], vm_kind: VMKind, config: &VMConfig) -> CryptoHash {
+    let hash = match CryptoHash::try_from(code_hash) {
+        Ok(hash) => hash,
+        // Sometimes caller doesn't compute code_hash, so hash the code ourselves.
+        Err(_e) => near_primitives::hash::hash(code),
+    };
+    let key = ContractCacheKey::Version1 {
+        code_hash: hash,
+        vm_config_non_crypto_hash: config.non_crypto_hash(),
+        vm_kind: vm_kind as u32,
+    };
+    near_primitives::hash::hash(&key.try_to_vec().unwrap())
+}
+
+fn cache_error(error: VMError, key: &CryptoHash, cache: &dyn CompiledContractCache) -> VMError {
+    let record = CacheRecord::Error(error.clone());
+    if cache.put(&(key.0).0, &record.try_to_vec().unwrap()).is_err() {
+        // That's fine, just cannot cache compilation error.
+        println!("Cannot cache an error");
+    }
+    error
+}
+
+fn compile_and_serialize_wasmer(
+    wasm_code: &[u8],
+    config: &VMConfig,
+    key: &CryptoHash,
+    cache: &dyn CompiledContractCache,
+) -> Result<wasmer_runtime::Module, VMError> {
+    let module = compile_module(wasm_code, config).map_err(|e| cache_error(e, &key, cache))?;
+    let artifact = module.cache().map_err(|_e| {
+        cache_error(
+            VMError::CacheError(SerializationError { hash: (key.0).0 }),
+            &key,
+            cache,
+        )
+    })?;
+    let code = artifact.serialize().map_err(|_e| {
+        VMError::CacheError(SerializationError { hash: (key.0).0 })
+    })?;
+    // If errors comes from serialization we shall not cache it.
+    let serialized = CacheRecord::Code(code).try_to_vec().map_err(|_e| {
+        VMError::CacheError(SerializationError { hash: (key.0).0 })
+    })?;
+    cache.put(key.as_ref(), &serialized).map_err(|_e| VMError::CacheError(WriteError))?;
+    Ok(module)
+}
+
+fn deserialize_wasmer(serialized: &[u8]) -> Result<wasmer_runtime::Module, VMError> {
+    let record = CacheRecord::try_from_slice(serialized)
+        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+    let code = match record {
+        CacheRecord::Error(err) => return Err(err),
+        CacheRecord::Code(code) => code,
+    };
+    let artifact = Artifact::deserialize(code.as_slice())
```

Comment:
Maybe rename it from `code` to `artifact`?

Reply:
Think the difference between two is that code is raw bytes, while artifact is Wasmer-specific structure.

Reply:
so it's a `serialized_artifact`